internal/plugin/grpc_provider.go

```diff
@@ -686,6 +693,93 @@ func (p *GRPCProvider) ReadDataSource(r providers.ReadDataSourceRequest) (resp p
 	return resp
 }
 
+func (p *GRPCProvider) CallFunction(r providers.CallFunctionRequest) (resp providers.CallFunctionResponse) {
+	logger.Trace("GRPCProvider: CallFunction")
+
+	schema := p.GetProviderSchema()
+	if schema.Diagnostics.HasErrors() {
+		// This should be unreachable
+		resp.Error = schema.Diagnostics.Err()
+		return resp
+	}
+
+	spec, ok := schema.Functions[r.Name]
+	if !ok {
+		// This should be unreachable
+		resp.Error = fmt.Errorf("invalid CallFunctionRequest: function %s not defined in provider schema", r.Name)
+		return resp
+	}
+
+	protoReq := &proto.CallFunction_Request{
+		Name:      r.Name,
+		Arguments: make([]*proto.DynamicValue, len(r.Arguments)),
+	}
+	if len(r.Arguments) < len(spec.Parameters) {
+		// This should be unreachable
+		resp.Error = fmt.Errorf("invalid CallFunctionRequest: function %s expected %d parameters and got %d instead", r.Name, len(spec.Parameters), len(r.Arguments))
+		return resp
+	}
+
+	// Translate the arguments
+	for i, arg := range r.Arguments {
+		// Most of the validation we do here should have already happened, but it does not hurt to be extra safe
```

Comment:
where do these validations happen? I didn't find them.

Reply:
Ah, that is assuming that cty/function.Function is calling this.  There's lots of internal logic there around handing parameters correctly and failing early if there are any issues.  I can clarify this comment!